### PR TITLE
SNAP-652 - Buckets with empty host-list are provided with other hosts.

### PR DESCRIPTION
**Pull request by ashetkar**
## Changes proposed in this pull request
- Partial fix for SNAP-652. Buckets with empty host-list (network servers) are given other buckets host-list for executors to process them on. See Other PRs section for the other part of the fix.

- Added a unit test for remote iteration in both the - unified and split cluster - modes.

## Patch testing
- precheckin
- New dunit test: ExternalShellDUnitTest.testRemoteIteratorSNAP652()

## Other PRs 
- https://github.com/SnappyDataInc/snappy-store/pull/52

### CODE REVIEW

**Comment by ashetkar** [comment]
@hbhanawat you too may want to take a look

**Comment by ashetkar** [comment]
@sumwale  you too may want to take a look at this.

File: core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala [hunk 1]
```diff
@@ -56,11 +56,14 @@ final class SparkShellRDDHelper {
     DriverRegistry.register(Constant.JDBC_CLIENT_DRIVER)
     val resolvedName = StoreUtils.lookupName(tableName, conn.getSchema)
     val par = split.index
-    val statement = conn.createStatement()
 
-    if (!useLocatorURL)
-      statement.execute(s"call sys.SET_BUCKETS_FOR_LOCAL_EXECUTION('$resolvedName', $par)")
+    if (!useLocatorURL) {
```
Comment (rishitesh) [inline core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala]: Any specific reason why this change ?

File: core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala [hunk 2]
```diff
@@ -149,6 +152,8 @@ object SparkShellRDDHelper {
         ClientAttribute.LOAD_BALANCE + "=false"
     val membersToNetServers = GemFireXDUtils.getGfxdAdvisor.
         getAllDRDAServersAndCorrespondingMemberMapping
+    var availableNetUrls = ArrayBuffer.empty[(String, String)]
```
Comment (rishitesh) [inline core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala]: it can be a val

File: core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala (hunk 1 above)
Comment (ashetkar) [inline core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala]: Intention was to have separate statement for each query to make it look neat. But I now realize that it is not supposed to be efficient.
I'll revert this specific change.

File: core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala (hunk 2 above)
Comment (ashetkar) [inline core/src/main/scala/io/snappydata/impl/SparkShellRDDHelper.scala]: Will change it.
